pkg/kube/cluster-init.sh

```diff
@@ -183,72 +338,210 @@ trigger_k3s_selfextraction() {
+echo "cluster-init.sh start for $HOSTNAME, uuid $DEVUUID" >> $INSTALL_LOG
+logmsg "Using ZFS persistent storage"
 
 #Forever loop every 15 secs
 while true;
 do
 if [ ! -f /var/lib/all_components_initialized ]; then
         if [ ! -f /var/lib/k3s_initialized ]; then
-                #/var/lib is where all kubernetes components get installed.
-                logmsg "Installing K3S version $K3S_VERSION"
+                logmsg "Installing K3S version $K3S_VERSION on $HOSTNAME"
                 mkdir -p /var/lib/k3s/bin
                 /usr/bin/curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=${K3S_VERSION} INSTALL_K3S_SKIP_ENABLE=true INSTALL_K3S_BIN_DIR=/var/lib/k3s/bin sh -
-                ln -s /var/lib/k3s/bin/* /usr/bin
+                sleep 5
                 logmsg "Initializing K3S version $K3S_VERSION"
+                ln -s /var/lib/k3s/bin/* /usr/bin
                 trigger_k3s_selfextraction
-                check_start_containerd
-                nohup /usr/bin/k3s server --config /etc/rancher/k3s/config.yaml &
-                #wait until k3s is ready
-                logmsg "Looping until k3s is ready"
-                while [ "$(kubectl get node "$(/bin/hostname)" -o json | jq '.status.conditions[] | select(.reason=="KubeletReady") | .status=="True"')" != "true" ];
-                do
-                        sleep 5;
-                done
-                # Give the embedded etcd in k3s priority over io as its fsync latencies are critical
-                ionice -c2 -n0 -p "$(pgrep -f "k3s server")"
-                logmsg "k3s is ready on this node"
-                # Default location where clients will look for config
-                ln -s /etc/rancher/k3s/k3s.yaml ~/.kube/config
                 touch /var/lib/k3s_initialized
         fi
 
-        if [ ! -f /var/lib/cni/bin ]; then
-                copy_cni_plugin_files
+        check_start_containerd
+        check_start_k3s
+
+        node_count_ready=$(kubectl get node | grep -w "$HOSTNAME" | grep -cw Ready )
+        if [ "$node_count_ready" -ne 1 ]; then
+                sleep 10
+                continue
+        fi
+        node_uuid_len=$(kubectl get nodes -l node-uuid="$DEVUUID" -o json | jq '.items | length')
+        if [ "$node_uuid_len" -eq 0 ]; then
+                logmsg "set node label with uuid $DEVUUID"
+                kubectl label node "$HOSTNAME" node-uuid="$DEVUUID"
         fi
+
         if [ ! -f /var/lib/multus_initialized ]; then
+                logmsg "Installing multus cni"
                 apply_multus_cni
+                sleep 10
+                continue
         fi
-        if ! pidof dhcp; then
-                # launch CNI dhcp service
-                /opt/cni/bin/dhcp daemon &
+
+        # setup debug user credential, role and binding
+        if [ ! -f /var/lib/debuguser-initialized ]; then
+                if are_all_pods_ready; then
+                        config_cluster_roles
+                fi
+                sleep 10
```

Comment:
There are a lot of 10s sleeps throughout sequential blocks of code... I'm afraid about the total waiting time, are those usually reached during initialization? Maybe a variable could be used instead of the hard coded 10 so it could be easily tuned when needed...

Reply:
These sleeps were also addressed in the broken up pr above, part 4 I believe.